Add ResNet-RS to keras.applications

spatil6 opened the pull request:
**Description**

This PR adds ResNet-RS model architecture to keras.applications. refer #15780 

Revisiting ResNets: Improved Training and Scaling Strategies

ResNet-RS models are updated versions of ResNet models - [Arxiv Link](https://arxiv.org/abs/2103.07579)
The models were rewritten using Keras functional API, The model's weights are converted from [original repository](https://github.com/tensorflow/tpu/tree/acb331c8878ce5a4124d4d7687df5fe0fadcd43b/models/official/resnet/resnet_rs).

**Motiviation**

Adding ResNetRS models to keras.applications has been discussed on Tensorflow's github:
#15780 
It was mentioned that this family of models would be welcomed in keras.applications, hence this contribution.

**Usage / Changes**

The models could be used similarly to other models in keras.applications

```
from keras.applications import resnet_rs
model = resnet_rs.ResNetRS50()
```
**Todos** 
Right now, weights are hosting on @sebastian-sz GitHub, need to move on keras-team gcs buckets.

haifeng-jin commented:
@LukeWood

haifeng-jin commented:
@qlzh727

fchollet commented on keras/applications/applications_load_weight_test.py:
```diff
@@ -71,6 +72,8 @@
         efficientnet_v2.EfficientNetV2L
     ]),
+    'resnet_rs':(resnet_rs,[resnet_rs.ResNetRS50,resnet_rs.ResNetRS101,resnet_rs.ResNetRS152,
+    resnet_rs.ResNetRS200,resnet_rs.ResNetRS270,resnet_rs.ResNetRS350,resnet_rs.ResNetRS420]),
```
Fix indent

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+
+def Conv2DFixedPadding(filters, kernel_size, strides, name):
+    """Conv2D block with fixed padding.
+    """
```
Move `"""` to line above

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    bn_momentum: float,
+    bn_epsilon: float,
+    activation: str,
+    se_ratio: float,
```
Most of these arguments should have default values

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    return apply
+
+
+def SE(in_filters: int, se_ratio: float, name: str, expand_ratio: int = 1):
```
`name` should go last and default to `None`. You can then use `get_uid` to uniquify the default name if None https://keras.io/api/utils/backend_utils/#getuid-function

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    }
+}
+
+def Conv2DFixedPadding(filters, kernel_size, strides, name):
```
See previous comment about `name`

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+        )(inputs)
+    return apply
+
+def STEM(bn_momentum: float, bn_epsilon: float, activation: str):
```
Inconsistent treatment of `name`: either you use it everywhere (preferred) or nowhere

spatil6 replied:
Added name.

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    num_repeats,
+    activation,
+    survival_probability: float,
+    name: str,
```
Same comment above `name`; most arguments above should have default values

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    input_shape=None,
+    input_tensor=None,
+    pooling=None,
+    classifier_activation="softmax",
```
We should also include `include_preprocessing=True` in the signature for all publicly exported applications

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+    include_preprocessing=True
+):
+    """Build Resnet-RS model, given provided parameters.
+    Parameters:
```
The docstring formatting should be fixed (see other docstrings in the Keras codebase for examples)

fchollet replied:
(Not yet done, please fix)

fchollet replied:
Another important note is that all these argument descriptions are redundant with the descriptions provided above. There should be a single source of truth so we don't have to copy/paste too much content when making modifications.

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+            x = layers.Rescaling(scale=1. / 255)(x)
+            x = layers.Normalization(
+                mean=[0.485, 0.456, 0.406],
+                variance=[0.229**2, 0.224**2, 0.225**2],
```
Why use mean-standardization here? In general `Rescaling` is enough for preprocessing. Subtracting the mean does not help the model.

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,891 @@
+                axis=bn_axis,
+            )(x)
+        else:
+            x = layers.Rescaling(scale=1. / 128.0, offset=-1)(x)
```
Why would the preprocessing logic be different in this case? This seems like a major gotcha

spatil6 replied:
It was mistake, forgot to remove earlier. Now taken care.

fchollet reviewed (changes requested):
Thanks for the PR -- sorry about the review delay!

LukeWood commented on keras/applications/applications_test.py:
```diff
@@ -76,7 +77,12 @@
                         (regnet.RegNetY080, 2016), (regnet.RegNetY120, 2240),
-                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712)]
+                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712),
+                        (resnet_rs.ResNetRS50,2048),(resnet_rs.ResNetRS101,2048),
```
Thanks for the PR and your patience.  Sorry, this fell through the cracks on my end for a bit.

add spaces after commas in line 82-84

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,919 @@
+    return apply
+
+
+def SE(in_filters: int, se_ratio: float =0.25,  expand_ratio: int = 1,name = None):
```
There are style issues in the code, e.g. spacing around operators. Could you please run `black` on the file?

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,919 @@
+    """Squeeze and Excitation block."""
+    bn_axis = 3 if backend.image_data_format() == "channels_last" else 1
+    if name is None:
+        name = "se" + str(backend.get_uid("se"))
```
Use a f-string and add a `_` before the counter

spatil6 replied:
Done

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,919 @@
+    bn_axis = 3 if backend.image_data_format() == "channels_last" else 1
+    x = img_input
+    if include_preprocessing:
+        x = layers.Rescaling(scale=1. / 255)(x)
```
Did you train the model from scratch to match this preprocessing scheme? Or were the weights extracted from somewhere?

sebastian-sz replied:
@fchollet The weights were converted by me, from the [original repository](https://github.com/tensorflow/tpu/tree/master/models/official/resnet/resnet_rs). The preprocessing layer should be:
```python
x = Normalization(
    mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
    variance=[(0.229 * 255) ** 2, (0.224 * 255) ** 2, (0.225 * 255) ** 2],
)(x)
``` 
to match the original preprocessing.

spatil6 replied:
Weights were converted by @sebastian-sz  👍 

I have also updated preprocessing layer.

Below you can find results for imagenet evaluation, 

<!DOCTYPE html>
Depth | Image size | Official Top 1 | With converted weights Top 1 | With converted weights Top 5 | Top 1 diff
-- | -- | -- | -- | -- | --
50 | 160 | 78.8 | 78.9 | 94.3 | +0.1
101 | 160 | 80.3 | 80.3 | 95.0 | 0.0
101 | 192 | 81.2 | 81.3 | 95.6 | +0.1
152 | 192 | 82.0 | 82.1 | 95.9 | +0.1
152 | 224 | 82.2 | 82.9 | 96.2 | +0.7
152 | 256 | 83.0 | 83.0 | 96.3 | 0.0
200 | 256 | 83.4 | 83.4 | 96.5 | 0.0
270 | 256 | 83.8 | 83.8 | 96.7 | 0.0
350 | 256 | 84.0 | 84.0 | 96.8 | 0.0
350 | 320 | 84.2 | 84.3 | 96.8 | +0.1
420 | 320 | 84.4 | 84.4 | 96.8 | 0.0

[Evaluation script](https://github.com/sebastian-sz/resnet-rs-keras/tree/main/imagenet_evaluation)

fchollet replied:
> Following are comparative imagenet evaluation results with official weights,

Thanks for the info! Did you generate these numbers from the latest code in this PR? Do you have a Colab I can check out?

spatil6 replied:
Those were earlier generated numbers.

Please refer below Colab link, where I have ran evaluation script on latest code from this PR.  Numbers are approximately matching with official repo.

https://colab.research.google.com/drive/1XoR18LoPhZ0Y1T3GXUdXvSmYlkJhSCS9?usp=sharing

fchollet commented on keras/applications/applications_test.py:
```diff
@@ -76,7 +77,12 @@
+                        (resnet_rs.ResNetRS50,2048), (resnet_rs.ResNetRS101,2048),
+                        (resnet_rs.ResNetRS152,2048), (resnet_rs.ResNetRS200,2048),
+                        (resnet_rs.ResNetRS270,2048), (resnet_rs.ResNetRS350,2048),
+                        (resnet_rs.ResNetRS420,2048)
```
In all added lines above, make sure to add a space after `,`

spatil6 replied:
Done.

fchollet commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,955 @@
+            use on the "top" layer. Ignored unless `include_top=True`. Set
+            `classifier_activation=None` to return the logits of the "top" layer.
+        include_preprocessing: Boolean, whether to include the preprocessing layer
+            (`Rescaling`) at the bottom of the network. Defaults to `True`.
```
This needs to be updated

spatil6 replied:
Updated.

spatil6 replied:
Done

fchollet reviewed (commented):
Thanks for the update!

spatil6 commented:
Thanks for review @fchollet . I have updated code according to review comment, also added Colab link of this PR code evaluation.

 https://colab.research.google.com/drive/1XoR18LoPhZ0Y1T3GXUdXvSmYlkJhSCS9?usp=sharing

fchollet reviewed (commented):
Thanks for the update!

spatil6 commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,959 @@
+            `classifier_activation=None` to return the logits of the "top" layer.
+        include_preprocessing: Boolean, whether to include the preprocessing layer
+            (`Rescaling`) at the bottom of the network. Defaults to `True`.
+            Note: if weights are 'imagenet' then Input image is normalized by ImageNet mean and standard deviation.
```
Done at both place.

fchollet reviewed (approved):
LGTM, thank you

LukeWood commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,959 @@
+    """Build Resnet-RS model, given provided parameters.
+
+    Args:
+        depth: Depth of ResNet network.
```
re-order to match the order in the function parameters

spatil6 replied:
Done.

spatil6 replied:
@LukeWood its done. Please let me know if any additional changes required

haifeng-jin commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,959 @@
+        dropout_rate: dropout rate before final classifier layer.
+        drop_connect_rate: dropout rate at skip connections.
+        include_top: whether to include the fully-connected layer at the top of
+        the network.
```
A indent is missing on this line.

spatil6 replied:
Done

LukeWood commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,957 @@
+):
+    """Build Resnet-RS model, given provided parameters.
+
+    Args:
```
unindent docstring to match code indentation

spatil6 replied:
Changed indentation from 4 spaces to 2 spaces, it should be good now. Let me know if anything, need to be done.

LukeWood commented on keras/applications/resnet_rs.py:
```diff
@@ -0,0 +1,957 @@
+            `classifier_activation=None` to return the logits of the "top" layer.
+        include_preprocessing: Boolean, whether to include the preprocessing layer
+            (`Rescaling`) at the bottom of the network. Defaults to `True`.
+            Note: Input image is normalized by ImageNet mean and standard deviation.
```
lets change "Note:" to "Note-".  "Note:" gets picked up as a parameter name by our docstring generator

spatil6 replied:
Done

LukeWood commented:
I've almost got this ready to merge.  Thanks for your patience!

fchollet commented:
This has been merged in https://github.com/keras-team/keras/commit/a9b4cb6bba4608d44f41f2719ec50356710cadb5 and https://github.com/tensorflow/tensorflow/commit/e9c5ec2787e3dbba0cc13aed554025e4c1c4edf1 -- thanks a lot for the great contribution! 👍

spatil6 commented:
> This has been merged in [a9b4cb6](https://github.com/keras-team/keras/commit/a9b4cb6bba4608d44f41f2719ec50356710cadb5) and [tensorflow/tensorflow@e9c5ec2](https://github.com/tensorflow/tensorflow/commit/e9c5ec2787e3dbba0cc13aed554025e4c1c4edf1) -- thanks a lot for the great contribution! +1

@fchollet should i open new PR to update keras-io docs for information about same?

sebastian-sz commented:
@shkarupa-alex 
Disclaimer: I am the author of the [Keras port repository](https://github.com/sebastian-sz/resnet-rs-keras).

Good find, I agree there is no mention of momentum in the paper and the value has been simply copied from the config file you linked. I missed that they used moving average decay. 

It could be beneficial to replace it with something more standard - e.g. maybe just leave the default batch norm momentum value?

I think you could open a separate issue regarding this problem. I don't think a closed PR will attract the needed attention.

LukeWood commented:
> @shkarupa-alex Disclaimer: I am the author of the [Keras port repository](https://github.com/sebastian-sz/resnet-rs-keras).
> 
> Good find, I agree there is no mention of momentum in the paper and the value has been simply copied from the config file you linked. I missed that they used moving average decay.
> 
> It could be beneficial to replace it with something more standard - e.g. maybe just leave the default batch norm momentum value?
> 
> I think you could open a separate issue regarding this problem. I don't think a closed PR will attract the needed attention.

agreed, lets file an issue.  maybe we can fix this with the `KerasCV.models` refactor?

sebastian-sz commented:
@LukeWood yes this is a minor change in code.

I think Resnet-RS models could be added to keras-cv with this fix. Do we also would like to modify `keras.applications` implementation?

LukeWood commented:
If it is a minor change I see no harm in also fixing keras.applicatins - as long as it doesn't change the API (but I don't think it does)

shkarupa-alex commented:
Here is a PR https://github.com/keras-team/keras/pull/16726

Pandaaaa906 commented:
will ResNet-RS merge into Keras V3?
